docs/advanced_examples/LogisticRegressionTraining.ipynb

```diff
@@ -462,6 +395,339 @@
+      "text/plain": [
+       "<Figure size 1000x600 with 1 Axes>"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "batch_size = sgd_clf_binary_fhe.batch_size\n",
+    "\n",
+    "# Initialize the weight and bias randomly\n",
+    "# They are going to be updated using FHE training.\n",
+    "weights = np.random.rand(1, 2, 1)\n",
+    "bias = np.random.rand(1, 1, 1)\n",
+    "\n",
+    "# Shuffle X_binary and y_binary\n",
+    "perm = np.random.permutation(X_binary.shape[0])\n",
+    "X_binary = X_binary[perm, ::]\n",
+    "y_binary = y_binary[perm]\n",
+    "\n",
+    "# Plot the decision boundaries before starting\n",
+    "plot_decision_boundary(\n",
+    "    X_binary,\n",
+    "    y_binary,\n",
+    "    weights=weights,\n",
+    "    bias=bias,\n",
+    "    title=\"Decision Boundary before training\",\n",
+    "    accuracy=compute_model_accuracy(weights, bias, X_binary, y_binary),\n",
+    ")\n",
+    "\n",
+    "\n",
+    "def train_fhe_client_server(\n",
+    "    X_binary,\n",
+    "    y_binary,\n",
+    "    batch_size,\n",
+    "    fhe_client,\n",
+    "    fhe_server,\n",
+    "    serialized_evaluation_key,\n",
+    "    weights,\n",
+    "    bias,\n",
+    "    n_epochs=1,\n",
+    "):\n",
+    "    acc_history = []\n",
+    "\n",
+    "    for epoch in range(n_epochs):\n",
+    "        # Shuffle X_binary and y_binary\n",
+    "        perm = np.random.permutation(X_binary.shape[0])\n",
+    "        X_binary = X_binary[perm, ::]\n",
+    "        y_binary = y_binary[perm]\n",
+    "\n",
+    "        for idx in range(X_binary.shape[0] // batch_size):\n",
+    "            batch_range = range(idx * batch_size, (idx + 1) * batch_size)\n",
+    "\n",
+    "            # Make the data X (1, batch_size, n_features)\n",
+    "            # and y (1, batch_size, n_targets)\n",
+    "            x_batch = X_binary[batch_range, :].reshape(1, -1, 2)\n",
+    "            y_batch = y_binary[batch_range].reshape(1, -1, 1)\n",
+    "\n",
+    "            # Encrypt the batch\n",
+    "            x_batch_enc = fhe_client.quantize_encrypt_serialize((x_batch, y_batch, weights, bias))\n",
+    "\n",
+    "            # Run the circuit on the server\n",
+    "            results = fhe_server.run(x_batch_enc, serialized_evaluation_key)\n",
+    "\n",
+    "            # Back on the client, we deserialize the result\n",
+    "            weights, bias = fhe_client.deserialize_decrypt_dequantize(results)\n",
+    "\n",
+    "            # Compute and store accuracy\n",
+    "            acc = compute_model_accuracy(weights, bias, X_binary, y_binary)\n",
+    "            acc_history.append(acc)\n",
+    "\n",
+    "        print(f\"Epoch {epoch + 1}/{n_epochs} completed. Accuracy: {acc_history[-1]}\")\n",
+    "\n",
+    "    return weights, bias, acc_history\n",
```

Comment:
haven't noticed that before, but I find it pretty weird that we are only handling weights and bias as results. The FheModelClient has the model's instance, wouldn't it be possible to return a trained SGDClassifier instead of the dequantized values ? 

because looking at the notebook I see that since we only have the weights, we need to re-implement the inference manually. Log reg is simple enough so that works but honestly not sure this is a good long-term solution

Reply:
So you are asking for a feature to be able to go back to SGDClassifier from FHEModelClient. This makes sense to me! But this requires more work. E.g. we will need to implement a method that load the learned parameter back in the model and every model will have it's own function so not obvious. I would rather implement that in another PR.

Reply:
https://github.com/zama-ai/concrete-ml-internal/issues/4464

Reply:
Yes agreed.